sound/soc/soc-pcm.c

```diff
@@ -2908,20 +2908,42 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
 	struct snd_pcm *pcm;
 	char new_name[64];
 	int ret = 0, playback = 0, capture = 0;
+	int stream;
 	int i;
 
+	if (rtd->dai_link->dynamic && rtd->num_cpus > 1) {
+		dev_err(rtd->dev,
+			"DPCM doesn't support Multi CPU for Front-Ends yet\n");
+		return -EINVAL;
+	}
+
 	if (rtd->dai_link->dynamic || rtd->dai_link->no_pcm) {
-		cpu_dai = asoc_rtd_to_cpu(rtd, 0);
-		if (rtd->num_cpus > 1) {
-			dev_err(rtd->dev,
-				"DPCM doesn't support Multi CPU yet\n");
-			return -EINVAL;
-		}
+		for_each_rtd_cpu_dais(rtd, i, cpu_dai) {
+
+			if (rtd->dai_link->dpcm_playback) {
+				stream = SNDRV_PCM_STREAM_PLAYBACK;
+				if (snd_soc_dai_stream_valid(cpu_dai, stream)) {
```

Comment:
Oh, ASoC makes my head hurt again. 

I don't quite get why "snd_soc_dai_stream_valid()==true" must apply for the first CPU DAI of the rtd, but for 2..N CPU DAIs, stream_valid check can fail.

Reply:
the idea was to check that ALL cpu_dais in a dailink that says it can do playack can indeed do so. So my proposal is to bail if you encounter one DAI that doesn't support playback.

Reply:
@plbossart I was thinking of following sequence:

CPU DAI #1 capture only but dai_link->dpcm_playback is set for the rtd so we go to L2923
CPU DAI #2 playback -> ok, playback set as 1
CPU DAI #3 another capture DAI -> as playback was set, we will return error that "CPU DAI #3 does not support playback"

... I guess this is not a case that can practically happen. If you think this is ok, I'm ok as well.

Reply:
no, you're right this doesn't work in all cases.
Maybe the correct thing to do is just increment a counter for valid dais, and check after the look that they are all valid.

Reply:
@kv2019i this check good to catch human errors with the dai drv array declaration in the driver.